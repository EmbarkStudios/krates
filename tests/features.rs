mod util;

/// Ensures weak dependencies are properly pruned if not explicitly pulled in
/// https://github.com/EmbarkStudios/krates/issues/41
#[test]
fn prunes_multiple_weak_features() {
    let mut cmd = krates::Cmd::new();
    cmd.manifest_path("tests/features/Cargo.toml")
        .features(
            ["blocking", "json", "multipart", "stream"]
                .into_iter()
                .map(|f| f.to_owned()),
        )
        .no_default_features();

    let mut builder = krates::Builder::new();
    builder.include_targets([("x86_64-unknown-linux-gnu", vec![])]);
    let md: krates::Krates = builder.build(cmd, krates::NoneFilter).unwrap();

    for name in md.krates().map(|k| &k.name) {
        println!("{name}");
    }

    // All of the specified features have a weak dependency on reqwest, so it
    // shouldn't be present in the graph
    assert_eq!(0, md.krates_by_name("reqwest").count());
    assert_eq!(0, md.krates_by_name("reqest").count());
}

/// While the zlib features brings in git2, the openssl dependency for both
/// git2 and git2-sys is optional and weak, and since we've not explicitly
/// enabled a feature to bring it in, it should not be present in the graph,
/// even though `cargo metadata` will list it in the graph
#[test]
fn prunes_mixed_dependencies() {
    let mut cmd = krates::Cmd::new();
    cmd.manifest_path("tests/features/Cargo.toml")
        .features(["zlib".to_owned()]);

    let mut builder = krates::Builder::new();
    builder.include_targets([("x86_64-unknown-linux-gnu", vec![])]);
    let md: krates::Krates = builder.build(cmd, krates::NoneFilter).unwrap();

    assert_eq!(0, md.krates_by_name("openssl-sys").count());
    // cmake is brought in via the zlib-ng-compat feature. gross.
    assert_eq!(1, md.krates_by_name("cmake").count());
}

macro_rules! assert_features {
    ($graph:expr, $name:expr, $features:expr) => {
        let (_, krate) = $graph.krates_by_name($name).next().unwrap();

        assert_eq!($graph.get_enabled_features(&krate.id).unwrap(), $features);
    };
}

/// Ensures we can enable crate features even when that crate has been renamed
/// as well as features being precise
#[test]
fn handles_features_for_renamed_crates() {
    let mut cmd = krates::Cmd::new();
    cmd.manifest_path("tests/features/Cargo.toml")
        .no_default_features()
        .features(["midi".to_owned()]);

    let mut builder = krates::Builder::new();
    builder.include_targets([("aarch64-apple-darwin", vec![])]);
    let md: krates::Krates = builder.build(cmd, krates::NoneFilter).unwrap();

    assert_eq!(1, md.krates_by_name("coreaudio-rs").count());

    // Ensure only the single feature is enabled, not all of them
    assert_features!(md, "coreaudio-sys", ["core_midi"]);
}

/// Ensures that explicitly toggling on an optional dependency works
#[test]
fn handles_explicit_weak_features() {
    {
        let mut cmd = krates::Cmd::new();
        cmd.manifest_path("tests/features/Cargo.toml")
            .no_default_features()
            .features(["reqest".to_owned()]);

        let mut builder = krates::Builder::new();
        builder.include_targets([("x86_64-unknown-linux-musl", vec![])]);
        let md: util::Graph = builder.build(cmd, krates::NoneFilter).unwrap();

        let dg = krates::petgraph::dot::Dot::new(md.graph()).to_string();

        std::fs::write("ohno.dot", dg).unwrap();

        // assert_features!(
        //     md,
        //     "reqwest",
        //     [
        //         "__rustls",
        //         "__tls",
        //         "async-compression",
        //         "brotli",
        //         "cookie_crate",
        //         "cookie_store",
        //         "cookies",
        //         "hyper-rustls",
        //         "proc-macro-hack",
        //         "rustls",
        //         "rustls-pemfile",
        //         "rustls-tls",
        //         "rustls-tls-webpki-roots",
        //         "tokio-rustls",
        //         "tokio-util",
        //         "webpki-roots"
        //     ]
        // );
    }

    // {
    //     let mut cmd = krates::Cmd::new();
    //     cmd.manifest_path("tests/features/Cargo.toml")
    //         .no_default_features()
    //         .features(["reqest".to_owned()]);

    //     let mut builder = krates::Builder::new();
    //     builder.include_targets([("x86_64-unknown-linux-musl", vec![])]);
    //     builder.ignore_kind(krates::DepKind::Normal, krates::Scope::All);
    //     let md: krates::Krates = builder.build(cmd, krates::NoneFilter).unwrap();

    //     assert_features!(
    //         md,
    //         "reqwest",
    //         ["cookie_crate", "cookie_store", "cookies", "proc-macro-hack"]
    //     );
    // }

    // let mut cmd = krates::Cmd::new();
    // cmd.manifest_path("tests/features/Cargo.toml")
    //     .features(["reqest".to_owned()]);

    // let mut builder = krates::Builder::new();
    // builder.include_targets([("x86_64-unknown-linux-musl", vec![])]);
    // let md: krates::Krates = builder.build(cmd, krates::NoneFilter).unwrap();

    // assert_features!(md, "reqwest", ["json", "cookies"]);
}
