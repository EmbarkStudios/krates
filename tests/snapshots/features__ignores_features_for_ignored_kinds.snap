---
source: tests/features.rs
expression: dotgraph
---
digraph {
    0 [ label = "crate ahash 0.7.7 (registry+https://github.com/rust-lang/crates.io-index)" ]
    1 [ label = "crate approx 0.4.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    2 [ label = "crate async-trait 0.1.77 (registry+https://github.com/rust-lang/crates.io-index)" ]
    3 [ label = "crate autocfg 1.1.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    4 [ label = "crate base64 0.13.1 (registry+https://github.com/rust-lang/crates.io-index)" ]
    5 [ label = "crate bitflags 1.3.2 (registry+https://github.com/rust-lang/crates.io-index)" ]
    6 [ label = "crate block-buffer 0.10.4 (registry+https://github.com/rust-lang/crates.io-index)" ]
    7 [ label = "crate cauchy 0.4.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    8 [ label = "crate cblas-sys 0.1.4 (registry+https://github.com/rust-lang/crates.io-index)" ]
    9 [ label = "crate cc 1.0.83 (registry+https://github.com/rust-lang/crates.io-index)" ]
    10 [ label = "crate cfg-if 1.0.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    11 [ label = "crate cmake 0.1.50 (registry+https://github.com/rust-lang/crates.io-index)" ]
    12 [ label = "crate config 0.13.4 (registry+https://github.com/rust-lang/crates.io-index)" ]
    13 [ label = "crate cpufeatures 0.2.12 (registry+https://github.com/rust-lang/crates.io-index)" ]
    14 [ label = "crate crypto-common 0.1.6 (registry+https://github.com/rust-lang/crates.io-index)" ]
    15 [ label = "crate digest 0.10.7 (registry+https://github.com/rust-lang/crates.io-index)" ]
    16 [ label = "crate dlv-list 0.3.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    17 [ label = "crate feature-bug 0.1.0 (path+file:///krates/tests/feature-bug)" ]
    18 [ label = "crate generic-array 0.14.7 (registry+https://github.com/rust-lang/crates.io-index)" ]
    19 [ label = "crate getrandom 0.2.12 (registry+https://github.com/rust-lang/crates.io-index)" ]
    20 [ label = "crate hashbrown 0.12.3 (registry+https://github.com/rust-lang/crates.io-index)" ]
    21 [ label = "crate itoa 1.0.10 (registry+https://github.com/rust-lang/crates.io-index)" ]
    22 [ label = "crate json5 0.4.1 (registry+https://github.com/rust-lang/crates.io-index)" ]
    23 [ label = "crate katexit 0.1.4 (registry+https://github.com/rust-lang/crates.io-index)" ]
    24 [ label = "crate lapack-sys 0.14.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    25 [ label = "crate lax 0.16.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    26 [ label = "crate lazy_static 1.4.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    27 [ label = "crate libc 0.2.152 (registry+https://github.com/rust-lang/crates.io-index)" ]
    28 [ label = "crate linked-hash-map 0.5.6 (registry+https://github.com/rust-lang/crates.io-index)" ]
    29 [ label = "crate matrixmultiply 0.3.8 (registry+https://github.com/rust-lang/crates.io-index)" ]
    30 [ label = "crate memchr 2.7.1 (registry+https://github.com/rust-lang/crates.io-index)" ]
    31 [ label = "crate minimal-lexical 0.2.1 (registry+https://github.com/rust-lang/crates.io-index)" ]
    32 [ label = "crate ndarray 0.15.6 (registry+https://github.com/rust-lang/crates.io-index)" ]
    33 [ label = "crate ndarray-linalg 0.16.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    34 [ label = "crate netlib-src 0.8.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    35 [ label = "crate nom 7.1.3 (registry+https://github.com/rust-lang/crates.io-index)" ]
    36 [ label = "crate num-complex 0.4.4 (registry+https://github.com/rust-lang/crates.io-index)" ]
    37 [ label = "crate num-integer 0.1.45 (registry+https://github.com/rust-lang/crates.io-index)" ]
    38 [ label = "crate num-traits 0.2.17 (registry+https://github.com/rust-lang/crates.io-index)" ]
    39 [ label = "crate once_cell 1.19.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    40 [ label = "crate ordered-multimap 0.4.3 (registry+https://github.com/rust-lang/crates.io-index)" ]
    41 [ label = "crate pathdiff 0.2.1 (registry+https://github.com/rust-lang/crates.io-index)" ]
    42 [ label = "crate pest 2.7.6 (registry+https://github.com/rust-lang/crates.io-index)" ]
    43 [ label = "crate pest_derive 2.7.6 (registry+https://github.com/rust-lang/crates.io-index)" ]
    44 [ label = "crate pest_generator 2.7.6 (registry+https://github.com/rust-lang/crates.io-index)" ]
    45 [ label = "crate pest_meta 2.7.6 (registry+https://github.com/rust-lang/crates.io-index)" ]
    46 [ label = "crate ppv-lite86 0.2.17 (registry+https://github.com/rust-lang/crates.io-index)" ]
    47 [ label = "crate proc-macro2 1.0.76 (registry+https://github.com/rust-lang/crates.io-index)" ]
    48 [ label = "crate quote 1.0.35 (registry+https://github.com/rust-lang/crates.io-index)" ]
    49 [ label = "crate rand 0.8.5 (registry+https://github.com/rust-lang/crates.io-index)" ]
    50 [ label = "crate rand_chacha 0.3.1 (registry+https://github.com/rust-lang/crates.io-index)" ]
    51 [ label = "crate rand_core 0.6.4 (registry+https://github.com/rust-lang/crates.io-index)" ]
    52 [ label = "crate rawpointer 0.2.1 (registry+https://github.com/rust-lang/crates.io-index)" ]
    53 [ label = "crate ron 0.7.1 (registry+https://github.com/rust-lang/crates.io-index)" ]
    54 [ label = "crate rust-ini 0.18.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    55 [ label = "crate ryu 1.0.16 (registry+https://github.com/rust-lang/crates.io-index)" ]
    56 [ label = "crate serde 1.0.195 (registry+https://github.com/rust-lang/crates.io-index)" ]
    57 [ label = "crate serde_derive 1.0.195 (registry+https://github.com/rust-lang/crates.io-index)" ]
    58 [ label = "crate serde_json 1.0.111 (registry+https://github.com/rust-lang/crates.io-index)" ]
    59 [ label = "crate sha2 0.10.8 (registry+https://github.com/rust-lang/crates.io-index)" ]
    60 [ label = "crate sub-crate 0.1.0 (path+file:///krates/tests/feature-bug/sub-crate)" ]
    61 [ label = "crate syn 1.0.109 (registry+https://github.com/rust-lang/crates.io-index)" ]
    62 [ label = "crate syn 2.0.48 (registry+https://github.com/rust-lang/crates.io-index)" ]
    63 [ label = "crate thiserror 1.0.56 (registry+https://github.com/rust-lang/crates.io-index)" ]
    64 [ label = "crate thiserror-impl 1.0.56 (registry+https://github.com/rust-lang/crates.io-index)" ]
    65 [ label = "crate toml 0.5.11 (registry+https://github.com/rust-lang/crates.io-index)" ]
    66 [ label = "crate typenum 1.17.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    67 [ label = "crate ucd-trie 0.1.6 (registry+https://github.com/rust-lang/crates.io-index)" ]
    68 [ label = "crate unicode-ident 1.0.12 (registry+https://github.com/rust-lang/crates.io-index)" ]
    69 [ label = "crate version_check 0.9.4 (registry+https://github.com/rust-lang/crates.io-index)" ]
    70 [ label = "crate wasi 0.11.0+wasi-snapshot-preview1 (registry+https://github.com/rust-lang/crates.io-index)" ]
    71 [ label = "crate yaml-rust 0.4.5 (registry+https://github.com/rust-lang/crates.io-index)" ]
    72 [ label = "feature alloc" ]
    73 [ label = "feature default" ]
    74 [ label = "feature default" ]
    75 [ label = "feature full" ]
    76 [ label = "feature visit-mut" ]
    77 [ label = "feature default" ]
    78 [ label = "feature serde" ]
    79 [ label = "feature rand" ]
    80 [ label = "feature default" ]
    81 [ label = "feature default" ]
    82 [ label = "feature default" ]
    83 [ label = "feature default" ]
    84 [ label = "feature default" ]
    85 [ label = "feature default" ]
    86 [ label = "feature default" ]
    87 [ label = "feature default" ]
    88 [ label = "feature default" ]
    89 [ label = "feature more_lengths" ]
    90 [ label = "feature simple" ]
    91 [ label = "feature default" ]
    92 [ label = "feature default" ]
    93 [ label = "feature full" ]
    94 [ label = "feature extra-traits" ]
    95 [ label = "feature default" ]
    96 [ label = "feature cblas" ]
    97 [ label = "feature static" ]
    98 [ label = "feature cgemm" ]
    99 [ label = "feature std" ]
    100 [ label = "feature std" ]
    101 [ label = "feature netlib" ]
    102 [ label = "feature blas" ]
    103 [ label = "feature approx" ]
    104 [ label = "feature std" ]
    105 [ label = "feature std" ]
    106 [ label = "feature std" ]
    107 [ label = "feature i128" ]
    108 [ label = "feature default" ]
    109 [ label = "feature default" ]
    110 [ label = "feature std" ]
    111 [ label = "feature std" ]
    112 [ label = "feature std" ]
    113 [ label = "feature default" ]
    114 [ label = "feature default" ]
    115 [ label = "feature proc-macro" ]
    116 [ label = "feature std" ]
    117 [ label = "feature alloc" ]
    118 [ label = "feature getrandom" ]
    119 [ label = "feature std" ]
    120 [ label = "feature simd" ]
    121 [ label = "feature std" ]
    122 [ label = "feature std" ]
    123 [ label = "feature default" ]
    124 [ label = "feature default" ]
    125 [ label = "feature serde_derive" ]
    126 [ label = "feature default" ]
    127 [ label = "feature std" ]
    128 [ label = "feature default" ]
    129 [ label = "feature default" ]
    130 [ label = "feature netlib" ]
    131 [ label = "feature default" ]
    132 [ label = "feature proc-macro" ]
    133 [ label = "feature std" ]
    134 [ label = "feature yaml-rust" ]
    135 [ label = "feature yaml" ]
    136 [ label = "feature toml" ]
    137 [ label = "feature serde_json" ]
    138 [ label = "feature rust-ini" ]
    139 [ label = "feature ron" ]
    140 [ label = "feature json5_rs" ]
    141 [ label = "feature json5_rs" ]
    142 [ label = "feature json5" ]
    143 [ label = "feature json" ]
    144 [ label = "feature ini" ]
    145 [ label = "feature core-api" ]
    146 [ label = "feature block-buffer" ]
    147 [ label = "feature inline-more" ]
    148 [ label = "feature ahash" ]
    149 [ label = "feature netlib-static" ]
    150 [ label = "feature std" ]
    151 [ label = "feature alloc" ]
    152 [ label = "feature libc" ]
    153 [ label = "feature cblas-sys" ]
    154 [ label = "feature std" ]
    155 [ label = "feature alloc" ]
    156 [ label = "feature std" ]
    157 [ label = "feature std" ]
    158 [ label = "feature race" ]
    159 [ label = "feature memchr" ]
    160 [ label = "feature std" ]
    161 [ label = "feature std_rng" ]
    162 [ label = "feature rand_chacha" ]
    163 [ label = "feature std" ]
    164 [ label = "feature alloc" ]
    165 [ label = "feature getrandom" ]
    166 [ label = "feature libc" ]
    167 [ label = "feature std" ]
    168 [ label = "feature quote" ]
    169 [ label = "feature proc-macro" ]
    170 [ label = "feature printing" ]
    171 [ label = "feature parsing" ]
    172 [ label = "feature derive" ]
    173 [ label = "feature clone-impls" ]
    174 [ label = "feature quote" ]
    175 [ label = "feature proc-macro" ]
    176 [ label = "feature printing" ]
    177 [ label = "feature parsing" ]
    178 [ label = "feature derive" ]
    179 [ label = "feature clone-impls" ]
    0 -> 19 [ label = " 'cfg(any(target_os = \"linux\", target_os = \"android\", target_os = \"windows\", target_os = \"macos\", target_os = \"ios\", target_os = \"freebsd\", target_os = \"openbsd\", target_os = \"netbsd\", target_os = \"dragonfly\", target_os = \"solaris\", target_os = \"illumos\", target_os = \"fuchsia\", target_os = \"redox\", target_os = \"cloudabi\", target_os = \"haiku\", target_os = \"vxworks\", target_os = \"emscripten\", target_os = \"wasi\"))'" ]
    0 -> 72 [ label = " 'cfg(not(all(target_arch = \"arm\", target_os = \"none\")))'" ]
    0 -> 69 [ label = "(build)" ]
    1 -> 38 [ label = "" ]
    2 -> 73 [ label = "" ]
    2 -> 74 [ label = "" ]
    2 -> 75 [ label = "" ]
    2 -> 76 [ label = "" ]
    2 -> 77 [ label = "" ]
    6 -> 18 [ label = "" ]
    7 -> 78 [ label = "" ]
    7 -> 79 [ label = "" ]
    7 -> 80 [ label = "" ]
    7 -> 81 [ label = "" ]
    7 -> 82 [ label = "" ]
    7 -> 83 [ label = "" ]
    8 -> 84 [ label = "" ]
    9 -> 27 [ label = " 'cfg(unix)'" ]
    11 -> 9 [ label = "" ]
    12 -> 2 [ label = "" ]
    12 -> 22 [ label = "" ]
    12 -> 26 [ label = "" ]
    12 -> 85 [ label = "" ]
    12 -> 41 [ label = "" ]
    12 -> 53 [ label = "" ]
    12 -> 86 [ label = "" ]
    12 -> 83 [ label = "" ]
    12 -> 87 [ label = "" ]
    12 -> 88 [ label = "" ]
    12 -> 71 [ label = "" ]
    13 -> 84 [ label = " 'aarch64-linux-android'" ]
    13 -> 84 [ label = " 'cfg(all(target_arch = \"aarch64\", target_os = \"linux\"))'" ]
    13 -> 84 [ label = " 'cfg(all(target_arch = \"aarch64\", target_vendor = \"apple\"))'" ]
    13 -> 84 [ label = " 'cfg(all(target_arch = \"loongarch64\", target_os = \"linux\"))'" ]
    14 -> 89 [ label = "" ]
    14 -> 66 [ label = "" ]
    15 -> 6 [ label = "" ]
    15 -> 14 [ label = "" ]
    17 -> 90 [ label = "" ]
    18 -> 66 [ label = "" ]
    18 -> 69 [ label = "(build)" ]
    19 -> 10 [ label = "" ]
    19 -> 27 [ label = " 'cfg(unix)'" ]
    19 -> 70 [ label = " 'cfg(target_os = \"wasi\")'" ]
    20 -> 0 [ label = "" ]
    22 -> 91 [ label = "" ]
    22 -> 92 [ label = "" ]
    22 -> 83 [ label = "" ]
    23 -> 73 [ label = "" ]
    23 -> 74 [ label = "" ]
    23 -> 93 [ label = "" ]
    23 -> 94 [ label = "" ]
    23 -> 95 [ label = "" ]
    24 -> 84 [ label = "" ]
    25 -> 7 [ label = "" ]
    25 -> 23 [ label = "" ]
    25 -> 24 [ label = "" ]
    25 -> 96 [ label = "" ]
    25 -> 97 [ label = "" ]
    25 -> 81 [ label = "" ]
    25 -> 63 [ label = "" ]
    29 -> 3 [ label = "(build)" ]
    29 -> 52 [ label = "" ]
    32 -> 1 [ label = "" ]
    32 -> 8 [ label = "" ]
    32 -> 84 [ label = "" ]
    32 -> 98 [ label = "" ]
    32 -> 99 [ label = "" ]
    32 -> 36 [ label = "" ]
    32 -> 37 [ label = "" ]
    32 -> 38 [ label = "" ]
    32 -> 100 [ label = "" ]
    32 -> 52 [ label = "" ]
    33 -> 7 [ label = "" ]
    33 -> 23 [ label = "" ]
    33 -> 25 [ label = "" ]
    33 -> 101 [ label = "" ]
    33 -> 102 [ label = "" ]
    33 -> 103 [ label = "" ]
    33 -> 104 [ label = "" ]
    33 -> 80 [ label = "" ]
    33 -> 81 [ label = "" ]
    33 -> 82 [ label = "" ]
    33 -> 63 [ label = "" ]
    34 -> 11 [ label = "(build)" ]
    35 -> 30 [ label = "" ]
    35 -> 105 [ label = "" ]
    35 -> 31 [ label = "" ]
    35 -> 106 [ label = "" ]
    36 -> 107 [ label = "" ]
    36 -> 100 [ label = "" ]
    36 -> 49 [ label = "" ]
    36 -> 56 [ label = "" ]
    37 -> 3 [ label = "(build)" ]
    37 -> 38 [ label = "" ]
    38 -> 3 [ label = "(build)" ]
    40 -> 16 [ label = "" ]
    40 -> 108 [ label = "" ]
    42 -> 109 [ label = "" ]
    42 -> 63 [ label = "" ]
    42 -> 67 [ label = "" ]
    42 -> 110 [ label = "" ]
    43 -> 42 [ label = "" ]
    43 -> 111 [ label = "" ]
    43 -> 44 [ label = "" ]
    43 -> 112 [ label = "" ]
    44 -> 42 [ label = "" ]
    44 -> 111 [ label = "" ]
    44 -> 113 [ label = "" ]
    44 -> 73 [ label = "" ]
    44 -> 74 [ label = "" ]
    44 -> 77 [ label = "" ]
    45 -> 114 [ label = "" ]
    45 -> 91 [ label = "" ]
    45 -> 59 [ label = "(build)" ]
    47 -> 68 [ label = "" ]
    48 -> 47 [ label = "" ]
    48 -> 115 [ label = "" ]
    49 -> 27 [ label = " 'cfg(unix)'" ]
    49 -> 50 [ label = "" ]
    49 -> 116 [ label = "" ]
    49 -> 51 [ label = "" ]
    49 -> 117 [ label = "" ]
    49 -> 118 [ label = "" ]
    49 -> 119 [ label = "" ]
    50 -> 120 [ label = "" ]
    50 -> 121 [ label = "" ]
    50 -> 51 [ label = "" ]
    51 -> 19 [ label = "" ]
    51 -> 122 [ label = "" ]
    53 -> 123 [ label = "" ]
    53 -> 124 [ label = "" ]
    53 -> 125 [ label = "" ]
    53 -> 83 [ label = "" ]
    54 -> 10 [ label = "" ]
    54 -> 40 [ label = "" ]
    56 -> 126 [ label = "" ]
    56 -> 126 [ label = " 'cfg(any())'" ]
    57 -> 73 [ label = "" ]
    57 -> 74 [ label = "" ]
    57 -> 77 [ label = "" ]
    58 -> 21 [ label = "" ]
    58 -> 55 [ label = "" ]
    58 -> 56 [ label = "" ]
    58 -> 127 [ label = "" ]
    59 -> 10 [ label = "" ]
    59 -> 13 [ label = " 'cfg(any(target_arch = \"aarch64\", target_arch = \"x86_64\", target_arch = \"x86\"))'" ]
    59 -> 128 [ label = "" ]
    60 -> 129 [ label = "" ]
    60 -> 130 [ label = " 'cfg(target_os = \"linux\")'" ]
    60 -> 131 [ label = " 'cfg(target_os = \"linux\")'" ]
    60 -> 130 [ label = " 'cfg(target_os = \"windows\")'" ]
    60 -> 131 [ label = " 'cfg(target_os = \"windows\")'" ]
    61 -> 47 [ label = "" ]
    61 -> 115 [ label = "" ]
    61 -> 48 [ label = "" ]
    61 -> 132 [ label = "" ]
    61 -> 68 [ label = "" ]
    62 -> 47 [ label = "" ]
    62 -> 115 [ label = "" ]
    62 -> 48 [ label = "" ]
    62 -> 132 [ label = "" ]
    62 -> 68 [ label = "" ]
    63 -> 64 [ label = "" ]
    64 -> 73 [ label = "" ]
    64 -> 74 [ label = "" ]
    64 -> 77 [ label = "" ]
    65 -> 83 [ label = "" ]
    71 -> 28 [ label = "" ]
    133 -> 4 [ label = "" ]
    123 -> 4 [ label = "" ]
    123 -> 133 [ label = "" ]
    124 -> 5 [ label = "" ]
    134 -> 12 [ label = "" ]
    134 -> 71 [ label = "" ]
    135 -> 12 [ label = "" ]
    135 -> 134 [ label = "" ]
    136 -> 12 [ label = "" ]
    136 -> 65 [ label = "" ]
    137 -> 12 [ label = "" ]
    137 -> 58 [ label = "" ]
    138 -> 12 [ label = "" ]
    138 -> 54 [ label = "" ]
    139 -> 12 [ label = "" ]
    139 -> 53 [ label = "" ]
    140 -> 12 [ label = "" ]
    140 -> 141 [ label = "" ]
    142 -> 12 [ label = "" ]
    142 -> 140 [ label = "" ]
    143 -> 12 [ label = "" ]
    143 -> 137 [ label = "" ]
    144 -> 12 [ label = "" ]
    144 -> 138 [ label = "" ]
    129 -> 12 [ label = "" ]
    129 -> 136 [ label = "" ]
    129 -> 143 [ label = "" ]
    129 -> 135 [ label = "" ]
    129 -> 144 [ label = "" ]
    129 -> 139 [ label = "" ]
    129 -> 142 [ label = "" ]
    128 -> 15 [ label = "" ]
    128 -> 145 [ label = "" ]
    145 -> 15 [ label = "" ]
    145 -> 146 [ label = "" ]
    146 -> 15 [ label = "" ]
    146 -> 6 [ label = "" ]
    89 -> 18 [ label = "" ]
    122 -> 19 [ label = "" ]
    147 -> 20 [ label = "" ]
    108 -> 20 [ label = "" ]
    108 -> 148 [ label = "" ]
    108 -> 147 [ label = "" ]
    148 -> 20 [ label = "" ]
    148 -> 0 [ label = "" ]
    149 -> 25 [ label = "" ]
    149 -> 97 [ label = "" ]
    101 -> 25 [ label = "" ]
    101 -> 149 [ label = "" ]
    150 -> 27 [ label = "" ]
    84 -> 27 [ label = "" ]
    84 -> 150 [ label = "" ]
    99 -> 29 [ label = "" ]
    98 -> 29 [ label = "" ]
    105 -> 30 [ label = "" ]
    105 -> 151 [ label = "" ]
    109 -> 30 [ label = "" ]
    109 -> 105 [ label = "" ]
    151 -> 30 [ label = "" ]
    106 -> 31 [ label = "" ]
    104 -> 32 [ label = "" ]
    104 -> 100 [ label = "" ]
    104 -> 99 [ label = "" ]
    152 -> 32 [ label = "" ]
    152 -> 27 [ label = "" ]
    153 -> 32 [ label = "" ]
    153 -> 8 [ label = "" ]
    102 -> 32 [ label = "" ]
    102 -> 153 [ label = "" ]
    102 -> 152 [ label = "" ]
    103 -> 32 [ label = "" ]
    103 -> 1 [ label = "" ]
    130 -> 33 [ label = "" ]
    130 -> 101 [ label = "" ]
    131 -> 33 [ label = "" ]
    97 -> 34 [ label = "" ]
    96 -> 34 [ label = "" ]
    154 -> 35 [ label = "" ]
    154 -> 155 [ label = "" ]
    154 -> 105 [ label = "" ]
    154 -> 106 [ label = "" ]
    85 -> 35 [ label = "" ]
    85 -> 154 [ label = "" ]
    155 -> 35 [ label = "" ]
    156 -> 36 [ label = "" ]
    156 -> 100 [ label = "" ]
    78 -> 36 [ label = "" ]
    78 -> 56 [ label = "" ]
    79 -> 36 [ label = "" ]
    79 -> 49 [ label = "" ]
    80 -> 36 [ label = "" ]
    80 -> 156 [ label = "" ]
    100 -> 38 [ label = "" ]
    107 -> 38 [ label = "" ]
    81 -> 38 [ label = "" ]
    81 -> 100 [ label = "" ]
    157 -> 39 [ label = "" ]
    157 -> 72 [ label = "" ]
    158 -> 39 [ label = "" ]
    114 -> 39 [ label = "" ]
    114 -> 157 [ label = "" ]
    72 -> 39 [ label = "" ]
    72 -> 158 [ label = "" ]
    111 -> 42 [ label = "" ]
    111 -> 110 [ label = "" ]
    111 -> 63 [ label = "" ]
    159 -> 42 [ label = "" ]
    159 -> 30 [ label = "" ]
    91 -> 42 [ label = "" ]
    91 -> 111 [ label = "" ]
    91 -> 159 [ label = "" ]
    160 -> 43 [ label = "" ]
    160 -> 111 [ label = "" ]
    160 -> 112 [ label = "" ]
    92 -> 43 [ label = "" ]
    92 -> 160 [ label = "" ]
    112 -> 44 [ label = "" ]
    112 -> 111 [ label = "" ]
    113 -> 45 [ label = "" ]
    121 -> 46 [ label = "" ]
    120 -> 46 [ label = "" ]
    115 -> 47 [ label = "" ]
    73 -> 47 [ label = "" ]
    73 -> 115 [ label = "" ]
    132 -> 48 [ label = "" ]
    132 -> 115 [ label = "" ]
    74 -> 48 [ label = "" ]
    74 -> 132 [ label = "" ]
    161 -> 49 [ label = "" ]
    161 -> 162 [ label = "" ]
    163 -> 49 [ label = "" ]
    163 -> 119 [ label = "" ]
    163 -> 116 [ label = "" ]
    163 -> 164 [ label = "" ]
    163 -> 165 [ label = "" ]
    163 -> 166 [ label = "" ]
    162 -> 49 [ label = "" ]
    162 -> 50 [ label = "" ]
    166 -> 49 [ label = "" ]
    166 -> 27 [ label = "" ]
    165 -> 49 [ label = "" ]
    165 -> 118 [ label = "" ]
    82 -> 49 [ label = "" ]
    82 -> 163 [ label = "" ]
    82 -> 161 [ label = "" ]
    164 -> 49 [ label = "" ]
    164 -> 117 [ label = "" ]
    116 -> 50 [ label = "" ]
    116 -> 121 [ label = "" ]
    119 -> 51 [ label = "" ]
    119 -> 117 [ label = "" ]
    119 -> 118 [ label = "" ]
    119 -> 122 [ label = "" ]
    118 -> 51 [ label = "" ]
    118 -> 19 [ label = "" ]
    117 -> 51 [ label = "" ]
    86 -> 54 [ label = "" ]
    127 -> 56 [ label = "" ]
    125 -> 56 [ label = "" ]
    125 -> 57 [ label = "" ]
    83 -> 56 [ label = "" ]
    83 -> 127 [ label = "" ]
    126 -> 57 [ label = "" ]
    167 -> 58 [ label = "" ]
    167 -> 127 [ label = "" ]
    87 -> 58 [ label = "" ]
    87 -> 167 [ label = "" ]
    90 -> 60 [ label = "" ]
    90 -> 33 [ label = "" ]
    168 -> 61 [ label = "" ]
    168 -> 48 [ label = "" ]
    169 -> 61 [ label = "" ]
    169 -> 115 [ label = "" ]
    169 -> 132 [ label = "" ]
    170 -> 61 [ label = "" ]
    170 -> 168 [ label = "" ]
    171 -> 61 [ label = "" ]
    93 -> 61 [ label = "" ]
    94 -> 61 [ label = "" ]
    172 -> 61 [ label = "" ]
    95 -> 61 [ label = "" ]
    95 -> 172 [ label = "" ]
    95 -> 171 [ label = "" ]
    95 -> 170 [ label = "" ]
    95 -> 173 [ label = "" ]
    95 -> 169 [ label = "" ]
    173 -> 61 [ label = "" ]
    76 -> 62 [ label = "" ]
    174 -> 62 [ label = "" ]
    174 -> 48 [ label = "" ]
    175 -> 62 [ label = "" ]
    175 -> 115 [ label = "" ]
    175 -> 132 [ label = "" ]
    176 -> 62 [ label = "" ]
    176 -> 174 [ label = "" ]
    177 -> 62 [ label = "" ]
    75 -> 62 [ label = "" ]
    178 -> 62 [ label = "" ]
    77 -> 62 [ label = "" ]
    77 -> 178 [ label = "" ]
    77 -> 177 [ label = "" ]
    77 -> 176 [ label = "" ]
    77 -> 179 [ label = "" ]
    77 -> 175 [ label = "" ]
    179 -> 62 [ label = "" ]
    88 -> 65 [ label = "" ]
    110 -> 67 [ label = "" ]
}

